src/vm/database/sqlite.rs

```diff
@@ -0,0 +1,305 @@
+use rusqlite::{Connection, OptionalExtension, NO_PARAMS, Row};
+use rusqlite::types::ToSql;
+
+use vm::contracts::Contract;
+use vm::database::{ContractDatabase};
+use vm::errors::{Error, ErrType, InterpreterResult as Result, IncomparableError};
+use vm::types::{Value, TypeSignature, TupleTypeSignature, AtomTypeIdentifier};
+
+
+pub struct SqliteContractDatabase {
+    conn: Option<Connection>,
+    save_point: u16
+}
+
+pub struct SqliteDataMap {
+    map_identifier: i64,
+    key_type: TypeSignature,
+    value_type: TypeSignature
+}
+
+impl SqliteContractDatabase {
+    pub fn initialize(filename: &str) -> Result<SqliteContractDatabase> {
+        let mut contract_db = SqliteContractDatabase::inner_open(filename)?;
+        contract_db.execute("CREATE TABLE IF NOT EXISTS maps_table
+                      (map_identifier INTEGER PRIMARY KEY AUTOINCREMENT,
+                       contract_name TEXT,
+                       map_name TEXT,
+                       key_type TEXT,
+                       value_type TEXT)",
+                            NO_PARAMS)?;
+        contract_db.execute("CREATE TABLE IF NOT EXISTS data_table
+                      (data_identifier INTEGER PRIMARY KEY AUTOINCREMENT,
+                       map_identifier INTEGER,
+                       key TEXT,
+                       value TEXT)",
+                            NO_PARAMS)?;
+        contract_db.execute("CREATE TABLE IF NOT EXISTS contracts
+                      (contract_identifier INTEGER PRIMARY KEY AUTOINCREMENT,
+                       contract_name TEXT,
+                       contract_data TEXT)",
+                            NO_PARAMS)?;
+
+        contract_db.check_schema()?;
+
+        Ok(contract_db)
+    }
+
+    pub fn open(filename: &str) -> Result<SqliteContractDatabase> {
+        let contract_db = SqliteContractDatabase::inner_open(filename)?;
+
+        contract_db.check_schema()?;
+        Ok(contract_db)
+    }
+
+    fn check_schema(&self) -> Result<()> {
```

Comment:
I'm not sure what this method does?  Is it just checking the structure of the db?

Reply:
Yeah, at the moment, it just checks that the tables exist, but once the schema finalizes, we'll want it to also check the schema (and potentially version check). The intent is to prevent the node from even starting up the VM on a bad database schema.

Reply:
Gotcha :+1: